File: pset6/vigenere.py
from sys import argv
from cs50 import get_string

keyLength = len(argv[1])


def main():

    # Check if the user has inputted the right number of arguments
    if len(argv) != 2:
        print("Usage: python vigenere.py key")
        exit(1)

    # Chekcs if the key contains any non-alphabetic characters
    for i in range(keyLength):
        if not argv[1][i].isalpha():
            print("Input should consist only of alphabetic characters")
            exit(1)

    # Get the plainyext from the user and print it
    plaintext = get_string("Plaintext: ")

    print(f"ciphertext: {encrypt(plaintext, argv[1])}",)

# Function to encrypt the plaintext


def encrypt(plaintext, key):
    ciphertext = ""
    j = 0
    # Iterates over characters in plaintext
    for i in plaintext:
        # Checks if character from plaintext is alphabetic, and if so proceed
        if i.isalpha():
            # Calculates the alphabetic value for the corresponding character for the key. And also rotates the characters using array indexing.
            if key[j % keyLength].isupper():
                keyValue = ord(key[j % keyLength]) - 65

            else:
                keyValue = ord(key[j % keyLength]) - 97

            # Calculates the value for the ciphertext character
            if i.isupper():
                ciphertext += chr(((ord(i) - 65 + keyValue) % 26) + 65)

            else:
                ciphertext += chr(((ord(i) - 97 + keyValue) % 26) + 97)

            j += 1

        # If the plaintext character isn't alphabetic just copy it to the ciphertext
        else:
            ciphertext += i

    return ciphertext


if __name__ == "__main__":
    main()

File: pset2/vigenere.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>
int shift(char c);
int main(int argc, string argv[])
{
    // Checks if the arguments are two, if not stop the program
    if (argc != 2)
    {
        printf("Usage: ./vigenere keyword\n");
        return 1;
    }
    // Assigns the length of the keyword to a variable for optimization purpouses, as it will be used multiple times
    int k = strlen(argv[1]);
    bool is_alpha = true;
    // A for loop to check if all the characters of the keyword are alphabetic
    for (int i = 0; i < k; i++)
    {
        if (!isalpha(argv[1][i]))
        {
            is_alpha = false;
        }
    }
    // If any of the keywords of the program aren't alphabetic, stop the program and return a message for correct use to the user
    if (is_alpha == false)
    {
        printf("Usage: ./vigenere keyword\n");
        return 1;
    }
    // Gets the plaintext from the user and stores it in a variable
    string plaintext = get_string("Plaintext: ");
    char key[strlen(plaintext)];
    char total[strlen(plaintext) + 1];
    int counter = 0;
    // Starts printing the ciphertext
    printf("Ciphertext: ");
    // Iterates over each character of the plaintext and adds the keyword to it, if the plaintext characters is alphabetic, else it just prints it
    for (int i = 0, j = strlen(plaintext); i < j; i++)
    {
        if (isalpha(plaintext[i]))
        {
            if (isupper(plaintext[i]))
            {
                key[i] = ((plaintext[i] - 'A') + (shift(argv[1][counter % k]))) % 26 + 'A';
            }
            else
            {
                key[i] = ((plaintext[i] - 'a') + (shift(argv[1][counter % k]))) % 26 + 'a';
            }
            counter++;
        }
        else
        {
            key[i] = plaintext[i];
        }
    }
    total[strlen(plaintext)] = '\0';
    // Prints a newline
    strcpy(total, key);
    printf("%s", total);
    printf("\n");

}
// Function to convert keyword character to number
int shift(char c)
{
    return toupper(c) - 'A';
}
